test/src/utils/differentiator.test.cpp

```diff
@@ -0,0 +1,328 @@
+/*
+ * Author: Donald Jennings
+ * Organisation: HYPED
+ * Date: 06/11/2020
+ * Description: Testing file for Differentiator.hpp
+ *
+ *    Copyright 2018 HYPED
+ *    Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ *    except in compliance with the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software distributed under
+ *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ *    either express or implied. See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <iostream>
+#include "math.h"
+#include "gtest/gtest.h"
+#include "utils/math/differentiator.hpp"
+#include "utils/system.hpp"
+#include "data/data_point.hpp"
+
+using hyped::data::DataPoint;
+using hyped::utils::math::Differentiator;
+
+// -------------------------------------------------------------------------------------------------
+// Functionality
+// -------------------------------------------------------------------------------------------------
+
+/**
+ * Struct used for test fixtures testing functionality.
+ * Constructor sets the values of the DataPoint objects.
+ * Helper functions defined to assist in testing
+ */
+struct DifferentiatorFunctionality : public ::testing::Test {
+ protected:
+  Differentiator<float> diff_test;
+  DataPoint<float> test_point;
+  DataPoint<float> second_point;
+  DataPoint<float> third_point;
+
+  void SetUp()
+  {
+    // Sets values of Data points
+    test_point.value = 10.5;
+    test_point.timestamp = 1;
+
+    second_point.value = 18.6;
+    second_point.timestamp = 2;
+
+    third_point.value= second_point.value;
+    third_point.timestamp = 3;
+  }
+
+  /**
+   * @brief Returns the difference between two data point values.
+  */
+  float deltaValue(DataPoint<float> point_1, DataPoint<float> prev_point)
```

Comment:
Usually, Structs do not contain function/methods so maybe you can leave them outside as helper functions

Reply:
What if you have a case where you cannot move a method outside (e.g. it uses the instance variables)? Normally, you switch from a `struct` to a `class` in these cases. :sweat_smile:

Reply:
Perhaps Classes could be more suited for building our Test Suites then since we can contain helper functions as methods within the class rather than making them all just global helper functions :)

Reply:
mmmm okay I agree that classes are better, however, I am still convinced that you could do it do everything with enums and extract the functions. If you need to use the instance variables you could pass that as an argument and call the function in set Up or even in the test (as it has also access to them). If you want to modify them you could pass it as a pointer. But yeah that is only me being headstrong.